src/main/clojure/clara/tools/inspect.clj

```diff
@@ -121,7 +124,17 @@
                          [rule
                           (for [token (mem/get-tokens-all memory rule-node)
                                 insertion (mem/get-insertions memory rule-node token)]
-                            {:explanation (first (to-explanations session [token])) :fact insertion})]))}))
+                            {:explanation (first (to-explanations session [token])) :fact insertion})]))
+
+     :fact->explanations (reduce (fn [previous [insertion info]]
+                                   (update previous insertion (fnil conj []) info))
+                                 {}
+                                 (mapcat (fn [[rule rule-node]]
+                                           (for [token (mem/get-tokens-all memory rule-node)
+                                                 insertion (mem/get-insertions memory rule-node token)]
+                                             [insertion {:rule rule
+                                                         :explanation (first (to-explanations session [token]))}]))
+                                         rule-to-nodes))}))
```

Comment:
Overall this looks good. Just one thing to consider: I'd be tempted to structure this a bit differently unless there are performance reasons not to. I think the above code could be accomplished like this:

```clj
     :fact->explanations (apply merge-with into
                                (for [[rule rule-node] rule-to-nodes
                                      token (mem/get-tokens-all memory rule-node)
                                      insertion (mem/get-insertions memory rule-node token)]

                                  {insertion [{:rule rule
                                               :explanation (first (to-explanations session [token]))}]}))
```

We can go with your implementation if this is too expensive or there is some other disadvantage. It just seems a bit more readable to me.